src/versions.rs

```diff
@@ -54,38 +54,63 @@ pub fn upgrade(mut pins_raw: Map<String, Value>) -> Result<Value> {
-                    "Pin {} could not be upgraded to the latest format version",
-                    name
-                ))?;
-            }
-        },
-        // All these versions are already handled by serde default fields
-        1 | 2 | 3 | 4 => {
-            log::info!("There is nothing to do");
-        },
-        5 => {
-            log::info!("sources.json is already up to date");
-        },
-        unknown => {
-            anyhow::bail!(
-                "Unknown format version {}, maybe try updating the application?",
-                unknown
-            );
-        },
+    /* A generic wrapper that updates all pins individually with a provided upgrade function.
+     * This can be used in all cases where only the pin structure and not the overall file structure
+     * changes, which should actually be most cases.
+     */
+    fn generic_upgrader(
+        pins_raw: &mut Map<String, Value>,
+        update_pin_fn: fn(&str, &mut Map<String, Value>) -> Result<()>,
+    ) -> Result<()> {
+        let pins = pins_raw
+            .get_mut("pins")
+            .and_then(Value::as_object_mut)
+            .ok_or_else(|| anyhow::format_err!("sources.json must contain a `pins` object"))?;
+        for (name, pin) in pins.iter_mut() {
+            update_pin_fn(
+                name,
+                pin.as_object_mut()
+                    .ok_or_else(|| anyhow::format_err!("Pin {} must be an object", name))?,
+            )
+            .context(anyhow::format_err!("Pin {} could not be upgraded", name))?;
+        }
+        Ok(())
+    }
+
+    /* Registry for version upgrade closures. Every uprade is registered for a version and will
+     * modify `pins_raw` to be of its following version.
+     * Most version upgrades are handled by serde default fields and don't need any special treatment.
+     * They are omitted here; Only non-trivial upgrades should be inserted.
+     */
+    type Upgrader = Box<dyn Fn(&mut Map<String, Value>) -> Result<()>>;
+    let version_upgraders: BTreeMap<u64, Upgrader> = [(
+        0,
+        Box::new(|pins_raw: &mut Map<String, Value>| generic_upgrader(pins_raw, upgrade_v0_pin))
+            as Upgrader,
+    )]
+    .into_iter()
+    .collect();
+
+    /* Some quick version checks to provide better user feedback */
+    if version > LATEST {
+        anyhow::bail!(
+            "Unknown format version {}, maybe try updating the application?",
+            version
+        );
+    } else if version == LATEST {
+        log::info!("sources.json is already up to date");
+    } else if version_upgraders.range(version..).next().is_none() {
+        /* We are not at the latest version, but there is no upgrader registered, which means that
+         * the upgrade is already being handled by serde default fields. The diff will always be
+         * the bump of the version field and nothing more.
+         */
+        log::info!("There is nothing to do");
```

Comment:
I feel like this test is not necessary, and could be replaced with
```rust
log::info("Upgrade complete");
```
after the `for` loop

Reply:
I migrated this code path out of inertia, but you're right that it isn't really needed. I'll remove it